adapters/mobkoi/mobkoi.go

```diff
@@ -119,3 +121,63 @@ func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.R
 	}
 	return bidResponse, nil
 }
+
+// This function returns the appropriate bidder endpoint, using the provided
+// integration endpoint if valid, otherwise falling back to the bidder endpoint from config.
+// Returns an error if no valid endpoint is available.
+func (a *adapter) getBidderEndpoint(bidderExt openrtb_ext.ImpExtMobkoi) (string, error) {
+	providedEndpoint := bidderExt.IntegrationEndpoint
+	if providedEndpoint != "" && isValidURL(providedEndpoint) {
+		return providedEndpoint, nil
+	}
+
+	if a.bidderEndpoint != "" && isValidURL(a.bidderEndpoint) {
+		return a.bidderEndpoint, nil
+	}
+
+	return "", fmt.Errorf("no valid endpoint configured: both integration endpoint (%s) and bidder endpoint (%s) are invalid", providedEndpoint, a.bidderEndpoint)
+}
+
+
+// This function checks if the endpoint is a valid URL.
+// Example valid and invalid URLs:
+// - https://adapter.config.bidder.endpoint.com/bid (valid)
+// - https://adapter.config.bidder.endpoint.com:8080/bid (valid)
+// - adapter.config.bidder.endpoint.com/bid (invalid)
+// - https://adapter.config.bidder.endpoint.com (invalid)
+func isValidURL(endpoint string) bool {
+	parsed, err := url.Parse(endpoint)
+	return err == nil && parsed.Scheme != "" && parsed.Host != "" && parsed.Path != "" && parsed.Path != "/"
+}
+
+// updateRequestExt sets the mobkoi extension fields in the request extension using standard JSON manipulation.
+func updateRequestExt(request *openrtb2.BidRequest) error {
+	// Parse existing request.Ext as map[string]json.RawMessage
+	extMap := make(map[string]json.RawMessage)
```

Comment:
Use `map[string]any` instead of `map[string]json.RawMessage` if modification is expected

`json.RawMessage` is useful when you want to defer parsing, but here you're immediately adding/modifying keys.